apisix/plugins/proxy-cache/memory_handler.lua

```diff
@@ -0,0 +1,317 @@
+        if err == "not found" then
+            return 404
+        end
+        ctx.cache.memory:purge(ctx.var.upstream_cache_key)
+        ctx.cache = nil
+        return 200
+    end
+
+    if err then
+        core.response.set_header("Apisix-Cache-Status", "MISS")
+        if err ~= "not found" then
+            core.log.error("failed to get from cache, err: ", err)
+        elseif conf.cache_control and cc["only-if-cached"] then
+            return 504
+        end
+        return
+    end
+
+    if res.version ~= CACHE_VERSION then
+        core.log.warn("cache format mismatch, purging ", ctx.var.upstream_cache_key)
+        core.response.set_header("Apisix-Cache-Status", "BYPASS")
+        ctx.cache.memory:purge(ctx.var.upstream_cache_key)
+        return
+    end
+
+    if conf.cache_control then
+        if cc["max-age"] and time() - res.timestamp > cc["max-age"] then
+            core.response.set_header("Apisix-Cache-Status", "STALE")
+            return
+        end
+
+        if cc["max-stale"] and time() - res.timestamp - res.ttl > cc["max-stale"] then
+            core.response.set_header("Apisix-Cache-Status", "STALE")
+            return
+        end
+
+        if cc["min-fresh"] and res.ttl - (time() - res.timestamp) < cc["min-fresh"] then
+            core.response.set_header("Apisix-Cache-Status", "STALE")
+            return
+        end
+    else
+        if time() - res.timestamp > res.ttl then
+            core.response.set_header("Apisix-Cache-Status", "STALE")
+            return
+        end
+    end
+
+    ctx.cache.hit = true
+
+    for key, value in pairs(res.headers) do
+        if conf.hide_cache_headers == true and include_cache_header(key) then
+            core.response.set_header(key, "")
+        elseif overwritable_header(key) then
+            core.response.set_header(key, value)
+        end
+    end
+
+    core.response.set_header("Age", floor(time() - res.timestamp))
+    core.response.set_header("Apisix-Cache-Status", "HIT")
+
+    return res.status, res.body
+end
+
+
+function _M.header_filter(conf, ctx)
+    local cache = ctx.cache
+    if not cache or cache.hit then
+        return
+    end
+
+    local res_headers = ngx.resp.get_headers(0, true)
+
+    for key in pairs(res_headers) do
+        if conf.hide_cache_headers == true and include_cache_header(key) then
+            core.response.set_header(key, "")
+        end
+    end
+
+    local cc = parse_directive_header(ctx.var.upstream_http_cache_control)
+
+    if cacheable_response(conf, ctx, cc) then
+        cache.res_headers = res_headers
+        cache.ttl = conf.cache_control and parse_resource_ttl(ctx, cc) or conf.cache_ttl
+    else
+        ctx.cache = nil
+    end
+end
+
+
+function _M.body_filter(conf, ctx)
+    local cache = ctx.cache
+    if not cache or cache.hit then
+        return
+    end
+
+    local chunk = ngx.arg[1]
+    local eof   = ngx.arg[2]
+
+    cache.res_body = (cache.res_body or "") .. (chunk or "")
```

Comment:
Could you use hold_body_chunk in https://github.com/apache/apisix/pull/5167 to optimize it?